File: modules/sekai_client_dummy/retriever.py
import asyncio
from typing import Optional
from base64 import b64decode

from ..logger import AsyncLogger
from ..schemas import InheritInformation, SekaiInheritDataRetrieverResponse
from ..enums import SupportedInheritUploadServer, UploadPolicy, UploadDataType
from .model import RequestData
from .client import ProjectSekaiClient
from .cryptor import pack, unpack

api = {SupportedInheritUploadServer.jp: "", SupportedInheritUploadServer.en: ""}
headers = {SupportedInheritUploadServer.jp: {}, SupportedInheritUploadServer.en: {}}
version = {SupportedInheritUploadServer.jp: "", SupportedInheritUploadServer.en: ""}


class ProjectSekaiDataRetriever:
    def __init__(
        self,
        server: SupportedInheritUploadServer,
        inherit: InheritInformation,
        policy: UploadPolicy,
        upload_type: UploadDataType,
        proxy: str = None,
    ) -> None:
        self.client = ProjectSekaiClient(
            server=server,
            api=api.get(server),
            version_info_url=version.get(server),
            inherit=inherit,
            headers=headers.get(server),
            pack_func=pack,
            unpack_func=unpack,
            proxy=proxy,
        )
        self.policy = policy
        self.upload_type = upload_type
        self.logger = AsyncLogger(__name__, level="DEBUG")
        self.is_error_exist = False
        self.error_message = None

    async def retrieve_suite(self) -> Optional[bytes]:
        if self.is_error_exist:
            return None
        await self.logger.info("Getting suite...")
        base_path = f""
        suite = await self.client.call_api(path=base_path)
        if suite:
            await asyncio.sleep(1)
            await self.logger.info("Calling suite...")
            path = base_path + ""
            await self.client.call_api(path=path)
            await asyncio.sleep(1)
            await self.client.call_api(path="/system")
            await asyncio.sleep(1)
            unpacked_suite = await self.client.unpack_data(suite)
            friend = bool(unpacked_suite["userFriends"])
            if self.client.login_bonus:
                if friend:
                    await asyncio.create_task(self.refresh_home(friends=True, login=True))
                else:
                    await asyncio.create_task(self.refresh_home(login=True))
            else:
                if friend:
                    await asyncio.create_task(self.refresh_home(friends=True))
                else:
                    await asyncio.create_task(self.refresh_home())
            return suite
        else:
            self.is_error_exist = True
            self.error_message = "Failed to retrieve suite, it may be due to API response timeout."
            return None

    async def refresh_home(self, friends: bool = False, login: bool = False) -> None:
        if self.is_error_exist:
            return None
        await self.logger.info("Refreshing home...")
        if friends:
            await self.client.call_api(path=f"")
            await self.client.call_api(path="/system")
            await self.client.call_api(path="/information")
        else:
            await self.client.call_api(path="/system")
            await self.client.call_api(path="/information")
        refresh_path = f""
        if login:
            return await self.client.call_api(
                path=refresh_path, method="PUT", data=await self.client.pack_data(RequestData.RefreshLogin)
            )
        else:
            return await self.client.call_api(
                path=refresh_path, method="PUT", data=await self.client.pack_data(RequestData.Refresh)
            )

    async def retrieve_mysekai(self) -> Optional[bytes]:
        if self.is_error_exist:
            return None
        response = await self.client.call_api(path="/module-maintenance/MYSEKAI")
        response = await self.client.unpack_data(response)
        if response["isOngoing"] is True:
            return None
        response = await self.client.call_api(path="/module-maintenance/MYSEKAI_ROOM")
        response = await self.client.unpack_data(response)
        if response["isOngoing"] is True:
            return None
        mysekai_data = await self.client.call_api(path=f"", method="POST", data=b64decode(RequestData.General))
        await self.client.call_api(path=f"", method="POST", data=await self.client.pack_data(RequestData.MySekaiRoom))
        await self.client.call_api(path=f"")
        return mysekai_data

    async def run(self) -> Optional[SekaiInheritDataRetrieverResponse]:
        await self.logger.start()
        await self.client.init()
        if self.client.is_error_exist:
            self.is_error_exist = True
            self.error_message = self.client.error_message
            await self.client.close()
            return None
        suite = await self.retrieve_suite()
        await self.refresh_home()
        if self.upload_type == UploadDataType.mysekai:
            mysekai = await self.retrieve_mysekai()
        else:
            mysekai = None
        await self.client.close()
        await self.logger.stop()
        return SekaiInheritDataRetrieverResponse(
            server=str(self.client.server),
            user_id=self.client.user_id,
            suite=suite,
            mysekai=mysekai,
            policy=self.policy,
        )

File: utils/sekai/retriever.go
package sekai

import (
	"context"
	"encoding/base64"
	"fmt"
	harukiUtils "haruki-suite/utils"
	harukiLogger "haruki-suite/utils/logger"
	"strconv"
	"strings"
	"time"
)

type DataRetriever struct {
	client       *HarukiSekaiClient
	uploadType   harukiUtils.UploadDataType
	logger       *harukiLogger.Logger
	isErrorExist bool
	ErrorMessage string
}

type HarukiSekaiDataRetriever = DataRetriever

func NewSekaiDataRetriever(
	server harukiUtils.SupportedInheritUploadServer,
	inherit harukiUtils.InheritInformation,
	uploadType harukiUtils.UploadDataType,
) *HarukiSekaiDataRetriever {
	client := NewSekaiClientWithConfig(ClientConfig{
		Server:          server,
		API:             Api[server],
		VersionURL:      Version[server],
		Inherit:         inherit,
		Headers:         Headers[server],
		Proxy:           thisProxy,
		InheritJWTToken: InheritJWTToken[server],
	})

	return &HarukiSekaiDataRetriever{
		client:       client,
		uploadType:   uploadType,
		logger:       harukiLogger.NewLogger("SekaiDataRetriever", "DEBUG", nil),
		isErrorExist: false,
		ErrorMessage: "",
	}
}

func (r *HarukiSekaiDataRetriever) RetrieveSuite(ctx context.Context) ([]byte, error) {
	if r.isErrorExist {
		return nil, NewDataRetrievalError("suite", "pre-check", r.ErrorMessage, nil)
	}
	serverName := strings.ToUpper(string(r.client.server))
	r.logger.Infof("%s server retrieving suite...", serverName)
	userIDStr := strconv.FormatInt(r.client.userID, 10)
	basePath := fmt.Sprintf("/suite/user/%s", userIDStr)
	suite, status, err := r.client.callAPI(ctx, basePath, "GET", nil, nil)
	if err != nil {
		r.logger.Errorf("Suite API call failed: %v", err)
		return nil, NewDataRetrievalError("suite", "api_call", "failed to call suite API", err)
	}
	if suite == nil {
		r.isErrorExist = true
		r.ErrorMessage = "suite API returned nil response"
		r.logger.Errorf("%s", r.ErrorMessage)
		return nil, NewDataRetrievalError("suite", "api_response", r.ErrorMessage, nil)
	}
	time.Sleep(1 * time.Second)
	r.logger.Debugf("%s server making follow-up suite calls...", serverName)
	path := basePath + "?isForceAllReload=false&name=user_colorful_pass,user_colorful_pass_v2,user_offline_event"
	if _, _, err := r.client.callAPI(ctx, path, "GET", nil, nil); err != nil {
		r.logger.Warnf("Follow-up suite call failed (non-critical): %v", err)
	}
	time.Sleep(1 * time.Second)
	if _, _, err := r.client.callAPI(ctx, "/system", "GET", nil, nil); err != nil {
		r.logger.Warnf("System call failed (non-critical): %v", err)
	}
	time.Sleep(1 * time.Second)
	unpacked, err := Unpack(suite, harukiUtils.SupportedDataUploadServer(r.client.server))
	if err != nil {
		r.logger.Errorf("Failed to unpack suite response: %v", err)
		return nil, NewDataRetrievalError("suite", "unpack", "failed to unpack response", err)
	}
	unpackedMap, ok := unpacked.(map[string]interface{})
	if !ok {
		r.logger.Errorf("Unexpected suite response type")
		return nil, NewDataRetrievalError("suite", "parse", "unexpected response type", nil)
	}
	hasFriends := false
	if f, ok := unpackedMap["userFriends"]; ok && f != nil {
		hasFriends = true
	}
	if err := r.RefreshHome(ctx, hasFriends, r.client.loginBonus); err != nil {
		r.logger.Warnf("RefreshHome failed (non-critical): %v", err)
	}
	if status == 200 {
		r.logger.Infof("%s server retrieved suite successfully.", serverName)
		return suite, nil
	}
	r.logger.Errorf("Suite API returned non-200 status: %d", status)
	return nil, NewDataRetrievalError("suite", "status", fmt.Sprintf("unexpected status code: %d", status), nil)
}

func (r *HarukiSekaiDataRetriever) RefreshHome(ctx context.Context, friends bool, login bool) error {
	if r.isErrorExist {
		return NewDataRetrievalError("home", "pre-check", r.ErrorMessage, nil)
	}
	serverName := strings.ToUpper(string(r.client.server))
	r.logger.Infof("%s server refreshing home...", serverName)
	userIDStr := strconv.FormatInt(r.client.userID, 10)
	var lastErr error
	if friends {
		if _, _, err := r.client.callAPI(ctx, fmt.Sprintf("/user/%s/invitation", userIDStr), "GET", nil, nil); err != nil {
			lastErr = err
			r.logger.Warnf("Invitation call failed: %v", err)
		}
	}
	if _, _, err := r.client.callAPI(ctx, "/system", "GET", nil, nil); err != nil {
		lastErr = err
		r.logger.Warnf("System call failed: %v", err)
	}
	if _, _, err := r.client.callAPI(ctx, "/information", "GET", nil, nil); err != nil {
		lastErr = err
		r.logger.Warnf("Information call failed: %v", err)
	}
	refreshPath := fmt.Sprintf("/user/%s/home/refresh", userIDStr)
	var refreshData map[string]interface{}
	if login {
		refreshData = RequestDataRefreshLogin
	} else {
		refreshData = RequestDataRefresh
	}
	data, err := Pack(refreshData, harukiUtils.SupportedDataUploadServer(r.client.server))
	if err != nil {
		r.logger.Warnf("Failed to pack refresh data: %v", err)
		return NewDataRetrievalError("home", "pack", "failed to pack refresh data", err)
	}
	if _, _, err := r.client.callAPI(ctx, refreshPath, "PUT", data, nil); err != nil {
		lastErr = err
		r.logger.Warnf("Home refresh call failed: %v", err)
	}
	r.logger.Infof("%s server home refresh completed.", serverName)
	if lastErr != nil {
		return NewDataRetrievalError("home", "refresh", "some refresh calls failed", lastErr)
	}
	return nil
}

func (r *HarukiSekaiDataRetriever) RetrieveMysekai(ctx context.Context) ([]byte, error) {
	if r.isErrorExist {
		return nil, NewDataRetrievalError("mysekai", "pre-check", r.ErrorMessage, nil)
	}
	serverName := strings.ToUpper(string(r.client.server))
	r.logger.Infof("%s server checking MySekai availability...", serverName)
	resp, status, err := r.client.callAPI(ctx, "/module-maintenance/MYSEKAI", "GET", nil, nil)
	if err != nil {
		r.logger.Warnf("MySekai maintenance check failed: %v", err)
		return nil, NewDataRetrievalError("mysekai", "maintenance_check", "failed to check maintenance status", err)
	}
	if status != 200 {
		return nil, NewDataRetrievalError("mysekai", "maintenance_check", fmt.Sprintf("unexpected status: %d", status), nil)
	}
	unpacked, err := Unpack(resp, harukiUtils.SupportedDataUploadServer(r.client.server))
	if err != nil {
		r.logger.Warnf("Failed to unpack maintenance response: %v", err)
	} else if m, ok := unpacked.(map[string]interface{}); ok && m["isOngoing"] == true {
		r.logger.Infof("MySekai is under maintenance")
		return nil, ErrMaintenance
	}
	resp, _, err = r.client.callAPI(ctx, "/module-maintenance/MYSEKAI_ROOM", "GET", nil, nil)
	if err != nil {
		r.logger.Warnf("MySekai Room maintenance check failed: %v", err)
	} else {
		unpacked, err = Unpack(resp, harukiUtils.SupportedDataUploadServer(r.client.server))
		if err != nil {
			r.logger.Warnf("Failed to unpack room maintenance response: %v", err)
		} else if m, ok := unpacked.(map[string]interface{}); ok && m["isOngoing"] == true {
			r.logger.Infof("MySekai Room is under maintenance")
			return nil, ErrMaintenance
		}
	}
	r.logger.Infof("%s server retrieving MySekai data...", serverName)
	userIDStr := strconv.FormatInt(r.client.userID, 10)
	general, err := base64.StdEncoding.DecodeString(RequestDataGeneral)
	if err != nil {
		return nil, NewDataRetrievalError("mysekai", "decode", "failed to decode request data", err)
	}
	mysekaiPath := fmt.Sprintf("/user/%s/mysekai?isForceAllReloadOnlyMySekai=True", userIDStr)
	mysekai, status, err := r.client.callAPI(ctx, mysekaiPath, "POST", general, nil)
	if err != nil {
		r.logger.Errorf("MySekai API call failed: %v", err)
		return nil, NewDataRetrievalError("mysekai", "api_call", "failed to call MySekai API", err)
	}
	roomReq, err := Pack(RequestDataMySekaiRoom, harukiUtils.SupportedDataUploadServer(r.client.server))
	if err != nil {
		r.logger.Warnf("Failed to pack room request: %v", err)
	} else {
		roomPath := fmt.Sprintf("/user/%s/mysekai/%s/room", userIDStr, userIDStr)
		if _, _, err := r.client.callAPI(ctx, roomPath, "POST", roomReq, nil); err != nil {
			r.logger.Warnf("Room call failed (non-critical): %v", err)
		}
	}
	diarkisPath := fmt.Sprintf("/user/%s/diarkis-auth?diarkisServerType=mysekai", userIDStr)
	if _, _, err := r.client.callAPI(ctx, diarkisPath, "GET", nil, nil); err != nil {
		r.logger.Warnf("Diarkis auth call failed (non-critical): %v", err)
	}
	if status == 200 {
		r.logger.Infof("%s server retrieved MySekai data successfully.", serverName)
		return mysekai, nil
	}
	r.logger.Errorf("MySekai API returned non-200 status: %d", status)
	return nil, NewDataRetrievalError("mysekai", "status", fmt.Sprintf("unexpected status code: %d", status), nil)
}

func (r *HarukiSekaiDataRetriever) Run(ctx context.Context) (*harukiUtils.SekaiInheritDataRetrieverResponse, error) {
	if err := r.client.Init(ctx); err != nil {
		r.isErrorExist = true
		r.ErrorMessage = err.Error()
		r.logger.Errorf("Client initialization failed: %v", err)
		return nil, NewDataRetrievalError("run", "init", "client initialization failed", err)
	}
	if r.client.isErrorExist {
		r.isErrorExist = true
		r.ErrorMessage = r.client.errorMessage
		r.logger.Errorf("Client error: %s", r.client.errorMessage)
		return nil, NewDataRetrievalError("run", "client_error", r.client.errorMessage, nil)
	}
	suite, suiteErr := r.RetrieveSuite(ctx)
	if suiteErr != nil {
		r.logger.Warnf("Suite retrieval failed: %v", suiteErr)
		// Continue to try other data retrieval
	}
	if err := r.RefreshHome(ctx, false, false); err != nil {
		r.logger.Warnf("Final home refresh failed (non-critical): %v", err)
	}
	var mysekai []byte
	var mysekaiErr error
	if r.uploadType == harukiUtils.UploadDataTypeMysekai {
		mysekai, mysekaiErr = r.RetrieveMysekai(ctx)
		if mysekaiErr != nil && !IsMaintenanceError(mysekaiErr) {
			r.logger.Warnf("MySekai retrieval failed: %v", mysekaiErr)
		}
	}
	return &harukiUtils.SekaiInheritDataRetrieverResponse{
		Server:  string(r.client.server),
		UserID:  r.client.userID,
		Suite:   suite,
		Mysekai: mysekai,
	}, nil
}
